package valhalla.odin;

message TripPath {

  message LatLng {
    optional float lat = 1;
    optional float lng = 2;
  }
 
  message Location {
    optional LatLng ll = 1;
    optional string name = 2;
    optional string phone = 3;
    optional string street = 4;    
    optional string city = 5;    
    optional string state = 6;        // TODO - others as per OSM; county?
    optional string postal_code = 7;    
    optional string country = 8;    
  }
  
  message Exit {                 // TODO - exit vs sign
    optional string number = 1;  // 67B
    repeated string branch = 2;  // I 95 N
    repeated string toward = 3;  // New York
    optional string name = 4;    // Not used much
  }
  
  enum Driveability {
    kNone = 0;
    kForward = 1;
    kBackward = 2;
    kBoth = 3;
  }
  
  message Edge {
    repeated string name = 1;               // street names
    optional float length = 2;              // km
    optional float speed = 3;               // km/h
    optional uint32 begin_heading = 4;      // 0-359
    optional uint32 end_heading = 5;        // 0-359
    optional uint32 begin_shape_index = 6;  // inclusive
    optional uint32 end_shape_index = 7;    // inclusive
    optional Driveability driveability = 8;
    optional bool ramp = 9;
    optional bool turn_channel = 10;
    optional bool ferry = 11;
    optional bool rail_ferry = 12;
    optional bool toll = 13;
    optional bool unpaved = 14;
    optional bool tunnel = 15;
    optional bool bridge = 16;
    optional bool roundabout = 17;
    optional bool internal_link = 18;
    optional uint32 end_node_index = 19;
    repeated Exit exit = 20;
  }
  
  message Node {
    optional uint64 id = 1;           // TODO - OpenLR?
    optional uint32 shape_index = 3;  //
    repeated Edge edge = 2;           // first one is the outbound path edge
  }
  
  optional uint64 trip_id = 1;
  optional uint32 leg_id = 2;
  repeated Location location = 3;
  repeated Node node = 4;
  optional string shape = 5;

}

